veps/sig-storage/dic-on-heterogeneous-cluster/dic-on-heterogeneous-cluster.md

```diff
@@ -0,0 +1,538 @@
+      deployed. It can be mitigated by not deploying a template, if its `DataSource` was not found.
+   
+   2. SSP can use the `cdi.kubevirt.io/dict.architectures` annotation in each `DataImportCronTemplate`, and create the
+      templates accordingly.
+   
+      The problem with this approach is that SSP will need to implement the logic to find out
+      which `DataImportCronTemplate` objects is related to each template. SSP can use the `spec.managedDataSource` field
+      in the `DataImportCronTemplate` object, to find the templates using the same `DataSource` name. It may be too 
+      complex to implement this solution.
+      > **Important**: this alternative will require no API changes in SSP, as the `spec.cluster.workloadArchitectures`
+        field is not needed.
+
+pros:
+
+- no changes are required from SSP, except for adopting the new CDI API, and changes regarding the VM templates.
+- The development of the changes in SSP and in HCO is not dependent on CDI.
+- If the 2nd option in the 2nd open question is chosen, the development of the changes in HCO is not dependent on SSP,
+  and each component can be developed and released independently.
+- In deployments without SSP, this solution will work as well, if the user creates the `DataImportCron` CRs manually.
+
+cons:
+
+- template handling in SSP seems to be more complex.
+
+### The Source of the Architecture List for Each Predefined Image
+For the predefined images, the architecture list is known in advance, and can be hardcoded in the DataImportCronTemplate
+file in the HCO image. HCO then will check what are the workload node architectures in the cluster, and will add the
+`ssp.kubevirt.io/dict.architectures` annotation to the `DataImportCronTemplate` object in the `SSP` CR, only with the
+architectures that are supported by the image, and by the cluster.
+
+However, these images are frequently updated, and we can never know if the architecture list that was set at HCO image
+build time, is still valid or not, at any point in the future.
+
+It seems like a low risk, but it will require a new release of HCO to update the architecture list.
+
+## Scalability
+
+n/a
+
+## Update/Rollback Compatibility
+
+### Upgrade Compatibility
+As outlined above, SSP should maintain the current behavior for `DataImportCronTemplate` objects, that
+do not include the `ssp.kubevirt.io/dict.architectures` annotation. Similarly, the cdi-importer preserve its existing
+behavior when the `DataSource`'s `spec.template.spec.source.registry.platform.architecture` is not populated.
+
+This ensures forward compatibility for all`DataImportCronTemplates` objects. For instance, even after
+upgrading the HCO to a version that implements this feature, the custom `DataImportCronTemplate` objects will remain
+unaltered and continue to function as expected, allowing users to update them at their own pace.
```

Comment:
From what I gather, the DataSource names for our default golden images change after this multi arch image support feature is turned on. That impacts user's existing investment in the default golden images.

Is there a path we can take that allows people to enable this feature while not making them do any "work" to adapt their current gitops flows simply to make their current yamls continue to work?

Reply:
Thanks for the important input, @davidvossel.

Could you please add more details? For the predefined (common) DICTs, and their corresponding templates, what manipulations did you see? I mean, do they change the DataSource name (or the `DATA_SOURCE_NAME` parameter)? how? can you please share an example?

Reply:
Not directly connected to David's comment, but we will have to change source DS names in templates too https://github.com/kubevirt/common-templates/issues/680

Reply:
Yes, but we still have to solve it for common templates itself. And wouldn't be better if the arch is already part of templates and then SSP wouldn't have to do anything with the `DATA_SOURCE_NAME`? I don't expect the DATA_SOURCE_NAME would change during the time.

Reply:
OK. I don't like using names for logic. I really prefer to use solid data like labels in this case. But I can see how setting naming standard can ease things. I can live with it...

Reply:
> Is there a path we can take that allows people to enable this feature while not making them do any "work" to adapt their current gitops flows simply to make their current yamls continue to work?

On a homogeneous cluster we could keep the DataSource names as is. On heterogeneous clusters though the name needs to be changed to avoid ambiguity.

Reply:
OK. So I  have a cluster and I decided to extand it. I've got atractive offer to purchase new machines, but with different architecture. So now my homogeneous cluster became heterogeneous. Should we keep the existing dataSources/DataImportCron/templates, or replace them with new arch specific ones?

Reply:
Can we make a DataSource arch-aware so it can point at multiple PVCs for different arches?

Reply:
I don't think so. I think @arnongilboa can explain better than I, but in general, IIUC, it requires meaningful not backward compatible apu changes.

Reply:
> So, with regards to seamless updates, what can be done to reduce disruption for people who are already using the existing DataSources by name?

So this is why I was hinting at making DataSources arch-aware, so the current names can be kept. But @aglitke is right that this is a mix of concerns.

A better alternative might be to keep a DataSource with the old name that points to an image with the same arch as the control plane architecture? This should allow for somewhat backward-compatible behavior?